### ISSUE TITLE
Very slow performance with simple kafka Consumer example

### ISSUE THREAD

**Issue by roikatz** [confluentinc/confluent-kafka-dotnet#554]
Description
===========
Hi,
We are using the KafkaConsumer based on the example from here: 
https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/examples/Consumer/Program.cs#L126

We are getting extremely slow performance (around 4000 messages per minute).
We can see that CPU utilisation was extremely low. (around 2%). seems like most of the time is waited to the consumer.

What can we do in order to speed up operation?
I can see differences between Java and .Net SDK. while in Java you get a list from Poll, in .Net you get message by message, either in Poll or consume APIs

How to reproduce
================

Code

   public class KafkaConsumer : IKafkaConsumer
    {
        private Consumer<string, string> consumer;
        private bool cancelled;
        private static ILog log = LogManager.GetLogger<KafkaConsumer>();
 
        public KafkaConsumer()
        {
            var kafkaBrokers = ConfigurationManager.AppSettings.Get("kafkaBrokerHost") ?? "localhost:9092";
            var kafkaConsumerGroupId = ConfigurationManager.AppSettings.Get("kafkaConsumerGroupId") ?? "dataplatform";
            var config = new Dictionary<string, object> { { "bootstrap.servers", kafkaBrokers }, { "group.id",  kafkaConsumerGroupId}, { "debug", "protocol" }, {"fetch.wait.max.ms","1" } };
            consumer = new Consumer<string, string>(config, new StringDeserializer(Encoding.UTF8), new StringDeserializer(Encoding.UTF8));
 
            // Note: All event handlers are called on the main thread.
            //consumer.OnLog += Loggers.ConsoleLogger;
            consumer.OnLog += (_, kafkaLog)
                =>
            {
                if (kafkaLog.Level < 7) log.Info($"Kafka log message: {kafkaLog}");
            };
 
            consumer.OnPartitionEOF += (_, end)
                => Console.WriteLine($"Reached end of topic {end.Topic} partition {end.Partition}, next message will be at offset {end.Offset}");
 
            consumer.OnError += (_, error)
                => Console.WriteLine($"Error: {error}");
 
            consumer.OnPartitionsAssigned += (_, partitions) =>
            {
                Console.WriteLine($"Assigned partitions: [{string.Join(", ", partitions)}], member id: {consumer.MemberId}");
                consumer.Assign(partitions);
            };
 
            consumer.OnPartitionsRevoked += (_, partitions) =>
            {
                Console.WriteLine($"Revoked partitions: [{string.Join(", ", partitions)}]");
                consumer.Unassign();
            };
 
            consumer.OnStatistics += (_, json)
                => Console.WriteLine($"Statistics: {json}");
        }
 
        public async Task Subscribe(Func<string, string, string, Task> processMessage, CancellationToken cancellationToken, params string[] topics)
        {
            cancelled = false;
            consumer.Subscribe(topics.ToList());
           
            while (!cancelled && !cancellationToken.IsCancellationRequested)
            {
                log.Debug("Before Consumer consume");
                Message<string, string> msg;
                if (!consumer.Consume(out msg, TimeSpan.FromMilliseconds(100)))
                {
                    log.Debug($"Consumer consume");
                    if (msg != null) log.Debug($"Kafka message error. key: {msg.Key} message: {msg.Error}");
                    continue;
                }
               
                Console.WriteLine($"Topic: {msg.Topic} Partition: {msg.Partition} Offset: {msg.Offset} {msg.Value}");
 
                log.Debug($"Topic: {msg.Topic} Partition: {msg.Partition} Offset: {msg.Offset} {msg.Key}");
                log.Debug("Before Start Proccessing");
                // Call the message processing callback
 
                //await processMessage(msg.Topic, msg.Key, msg.Value);
                log.Debug("After Start Proccessing");
                // Manually commit offsets
                if (msg.Offset % 5 == 0)
                {
                    Console.WriteLine($"Committing offset");
                    var committedOffsets = await consumer.CommitAsync(msg);
                    Console.WriteLine($"Committed offset: {committedOffsets}");
                    log.Debug($"commit finished for message {msg.Key}");
                }
            }
        }
 
        public void Unsubscribe()
        {
            cancelled = true;
            consumer.Unsubscribe();
        }
    }
}

Checklist
=========
Please provide the following information:

 - [ ] Confluent.Kafka nuget version: 0.11.0.0
 - [ ] Apache Kafka version: confluent-kafka-2.11-1.0.1-1
 - [ ] Client configuration: .net 4.5.2 
 - [ ] Operating system:  Kafka Centos 7.4, Client on Windows Server 2016 1607
 - [ ] Provide logs (with "debug" : "..." as necessary in configuration)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
librdkafka 0.11.0 (which confluent-kafka-dotnet 0.11.0 depends on) had some severe performance issues.
Strongly suggest you upgrade to Confluent-kafka-dotnet 0.11.4

**Comment by roikatz** [comment]
it does seems like it has a nice positive impact on performance, however, is there an example of how to use kafka with dotnet with high throughput configuration/code?

**Comment by mhowlett** [comment]
https://www.confluent.io/white-paper/optimizing-your-apache-kafka-deployment/

**Comment by mhowlett** [comment]
Have a look at: https://github.com/confluentinc/confluent-kafka-dotnet/tree/master/test/Confluent.Kafka.Benchmark

and also: https://www.confluent.io/white-paper/optimizing-your-apache-kafka-deployment/

the default parameters are good for high throughput, except I'd suggest increasing `linger.ms` (to 5 or 10 should be enough) to ensure good batching.

**Comment by mhowlett** [comment]
also, producing synchronously (waiting for a delivery report of a produce request before sending the next) or committing offsets for each consumed message will kill throughput. 

also, if you care about latency and you're running on windows, have a read or #501

**Comment by roikatz** [comment]
Thanks, will be checked.

**Comment by roikatz** [comment]
Hi again,
so with the new SDK, we are able to draw only around 1000 records per second, however CPU of the consumer (and KAFKA servers) looks very IDLE.

any ideas?

**Comment by mhowlett** [comment]
Awaiting `commit.async` is bound by round-trip latency to the cluster. About the fastest you could expect over a fast network is ~5ms. Since commit is being called every 5 messages, back of the envelope calculation gives 1000s/5ms * 5 = 1000 messages expected consume throughput. For higher throughput either commit less frequently, or keep the `enable.auto.commit` set to `true`. Actually, I notice you're not setting this, so in addition to manually committing, librdkafka is auto committing offsets in a background thread. Something else you could do is disable `enable.auto.offset.store` whilst enabling `enable.auto.commit` and use `Consumer.StoreOffsets`.

**Comment by mhowlett** [comment]
you could also try setting `socket.blocking.max.ms` to `1` (i'm unsure if it applies in this situation or not without investigating). see #574

**Comment by roikatz** [comment]
Hi,
Sorry for the long overdue, we still feeling the slowness with that SDK, i'm sure we set it up wrong.
Upgraded to 0.11.5, remove the await from the commit, trying to have the socket.blocking.max.ms to be 1 or just default, committed after 400, 4000, 5 and 100. It seems like nothing is making the consumption faster, We also changed the consume interface to the OnMessage event.
I'm kinda clueless here.

BTW, for our producer socket.blocking.max.ms=1 indeed improved the latency dramatically.
